stubs/CoreGenericFunctions.phpstub

```diff
@@ -1057,6 +1057,33 @@ function str_shuffle(string $string): string {}
  */
 function str_split(string $string, int $length = 1) {}
 
+/**
+ * @psalm-pure
+ * @template T as string
+ * @param T $needle
+ * @psalm-assert-if-true =(T is '' ? string : non-empty-string) $haystack
```

Comment:
Actually this should be written as
`@psalm-assert-if-true (T is '' ? string : =non-empty-string) $haystack`
but it explodes psalm parser.
But it's a really edge case when this difference matters. I don't think there would be any issues against any real-world code.